### ISSUE TITLE
Support VoodooI2C for touchpad

### ISSUE THREAD

**Issue by sunrez** [tylernguyen/x1c6-hackintosh#13]
Hi Tyler,

I notice that you're updating your repo's with latest kexts, etc.  Can you take a look at VoodooI2C for the touchpad?  This should enable gestures for the trackpad.

**Comment by tylernguyen** [comment]
Hi @sunrez,
Yes, that's one of current to-do right now. There's still a lot to do with the keyboard specifically also. I need to do a full mapping of the keyboard and its layers, then assign each key again properly (mostly the function and F keys). However, I do have a full time commitment so I cannot give you an eta.

**Comment by sunrez** [comment]
Hi @tylernguyen ,

Understood on timeframe, this stuff is obviously secondary to full-time commitments. 

**Question for you:**   Do you currently have two finger, three finger and four-finger gestures working with the PS2 kext for the touchpad?

Also as you probably know the Gen7 just came out.  If you have any details write-ups on the DSDT/ACPI stuff you're having to do, that will definitely help with Gen7 work.  I'm debating picking up a X1C7 and basing it off your work, but I too have full-time commitments.

**Comment by velaar** [comment]
@sunrez

I believe soldered Wifi will be an issue for X1C7

**Comment by AutinMitra** [comment]
I don't believe VoodooPS2 will work with the X1C6, as the trackpad uses PS2 instead of a I2C controller. If you want fully supported gestures, acidanthera has a fork of VoodooPS2, which emulates the Magic Trackpad 2 (just like VoodooI2C), and borrows some code from VoodooI2C.

Check it out: [VoodooPS2](https://github.com/acidanthera/VoodooPS2)

You need to compile the latest version, at least that's what I had to do

**Comment by sunrez** [comment]
Hi @velaar ,

Possibly, but looking at the X1C7 teardowns, the LTE/WWAN slot is open which means a BRCM M.2 wifi card could put in.  I don't know if the soldered down WiFi can be permanently turned off.

@AutinMitra ,

Are you using this multi-touch build of VoodooPS2 on the X1C6?  If so, can you talk about your experience with it?

**Comment by AutinMitra** [comment]
@sunrez 
It's actually been a pretty decent experience, the gestures and scroll are somewhat jittery (but feels better than standard VoodooPS2). It lacks palm rejection (as VoodooI2C is, but it may not be too hard to implement since the finger width is returned in the driver). There may be some bugs, as its a WIP, but I'm using it without too many issues. I'm currently running Catalina.

Also, for the X1C7, in BIOS I'm pretty sure you can disable WLAN (Wifi) and enable the WWAN. I'm not sure if toggling the WWAN option will enable the WWAN slot or the sim card slot, maybe ask an owner from the Lenovo community if that is the case?

**Comment by notthebee** [comment]
X1C6 has a Synaptics trackpad that only works in PS2 mode on macOS.
However, after some digging it seems that ELAN trackpad from T480s should be physically compatible with X1C6. This trackpad supports I2C mode with all the fancy gestures via this driver: https://github.com/leo-labs/VoodooSMBus
The tradeoff is the mylar coating which doesn't feel as nice as the glass trackpad on X1C6

I ordered the FRU on Aliexpress, will report back as soon as it arrives 🙂

**Comment by AutinMitra** [comment]
> X1C6 has a Synaptics trackpad that only works in PS2 mode on macOS.
> However, after some digging it seems that ELAN trackpad from T480s should be physically compatible with X1C6. This trackpad supports I2C mode with all the fancy gestures via this driver: https://github.com/leo-labs/VoodooSMBus
> The tradeoff is the mylar coating which doesn't feel as nice as the glass trackpad on X1C6
> 
> I ordered the FRU on Aliexpress, will report back as soon as it arrives

If you want the fancy gestures, there's no need to buy another touchpad. Acidanthera's VoodooPS2 had full gestures on PS2 on V2.0.4 (2.1.0 is broken on X1C6). Am using it on my X1C6, gestures works exactly like a MacBook Pro.

**Comment by velaar** [comment]
> PS2 with V2.0.4 (2.1.0 is broken on X1C6)
Running X1C6 with 2.1.0

```velaar@master ~ $ kextstat  | grep -i voodoo
  172    3 0xffffff7f847c9000 0x10000    0x10000    as.acidanthera.voodoo.driver.PS2Controller (2.1.0) 3080C899-33C9-3589-8DE4-9CDA254B4E84 <12 8 6 5 3 1>
  178    0 0xffffff7f847d9000 0x6000     0x6000     as.acidanthera.voodoo.driver.PS2Keyboard (2.1.0) 355D67D0-0559-39DA-8DCD-61A306CB7942 <172 63 8 6 5 3 1>
  179    0 0xffffff7f847df000 0xb000     0xb000     as.acidanthera.voodoo.driver.PS2Trackpad (2.1.0) C2C21C79-4A90-331A-A861-863E4C31ED50 <172 63 6 5 3>
  180    0 0xffffff7f847ea000 0x4000     0x4000     as.acidanthera.voodoo.driver.PS2Mouse (2.1.0) CC0B95F8-B53B-3F84-B341-E03FF6312F3C <172 63 6 5 3>
  189    0 0xffffff7f847f3000 0x6000     0x6000     me.kishorprins.VoodooInput (1) 00497A66-F6A7-3431-846D-2A43361C7C05 <63 6 5 3>```

No issues.

**Comment by AutinMitra** [comment]
> > PS2 with V2.0.4 (2.1.0 is broken on X1C6)
> 
> Running X1C6 (20KH) with 2.1.0 without any issues
> 
> ```
>   172    3 0xffffff7f847c9000 0x10000    0x10000    as.acidanthera.voodoo.driver.PS2Controller (2.1.0) 3080C899-33C9-3589-8DE4-9CDA254B4E84 <12 8 6 5 3 1>
>   178    0 0xffffff7f847d9000 0x6000     0x6000     as.acidanthera.voodoo.driver.PS2Keyboard (2.1.0) 355D67D0-0559-39DA-8DCD-61A306CB7942 <172 63 8 6 5 3 1>
>   179    0 0xffffff7f847df000 0xb000     0xb000     as.acidanthera.voodoo.driver.PS2Trackpad (2.1.0) C2C21C79-4A90-331A-A861-863E4C31ED50 <172 63 6 5 3>
>   180    0 0xffffff7f847ea000 0x4000     0x4000     as.acidanthera.voodoo.driver.PS2Mouse (2.1.0) CC0B95F8-B53B-3F84-B341-E03FF6312F3C <172 63 6 5 3>
>   189    0 0xffffff7f847f3000 0x6000     0x6000     me.kishorprins.VoodooInput (1) 00497A66-F6A7-3431-846D-2A43361C7C05 <63 6 5 3>```
> ```

What version are you running? I am on 10.15.1, and every trackpad input needs one more input than it should (two finger for moving, three for scrolling, etc). Here's a bugtracker where other people had similar issues: [Issue](https://github.com/acidanthera/bugtracker/issues/600)

**Comment by velaar** [comment]
> > > PS2 with V2.0.4 (2.1.0 is broken on X1C6)
> > 
> > 
> > Running X1C6 (20KH) with 2.1.0 without any issues
> > ```
> >   172    3 0xffffff7f847c9000 0x10000    0x10000    as.acidanthera.voodoo.driver.PS2Controller (2.1.0) 3080C899-33C9-3589-8DE4-9CDA254B4E84 <12 8 6 5 3 1>
> >   178    0 0xffffff7f847d9000 0x6000     0x6000     as.acidanthera.voodoo.driver.PS2Keyboard (2.1.0) 355D67D0-0559-39DA-8DCD-61A306CB7942 <172 63 8 6 5 3 1>
> >   179    0 0xffffff7f847df000 0xb000     0xb000     as.acidanthera.voodoo.driver.PS2Trackpad (2.1.0) C2C21C79-4A90-331A-A861-863E4C31ED50 <172 63 6 5 3>
> >   180    0 0xffffff7f847ea000 0x4000     0x4000     as.acidanthera.voodoo.driver.PS2Mouse (2.1.0) CC0B95F8-B53B-3F84-B341-E03FF6312F3C <172 63 6 5 3>
> >   189    0 0xffffff7f847f3000 0x6000     0x6000     me.kishorprins.VoodooInput (1) 00497A66-F6A7-3431-846D-2A43361C7C05 <63 6 5 3>```
> > ```
> 
> What version are you running? I am on 10.15.1, and every trackpad input needs one more input than it should (two finger for moving, three for scrolling, etc). Here's a bugtracker where other people had similar issues to me: [Issue](https://github.com/acidanthera/bugtracker/issues/600)

![image](https://user-images.githubusercontent.com/4062230/70392488-98fddf80-19ae-11ea-85f4-f5a6a8300acc.png)

**Comment by notthebee** [comment]
> > X1C6 has a Synaptics trackpad that only works in PS2 mode on macOS.
> > However, after some digging it seems that ELAN trackpad from T480s should be physically compatible with X1C6. This trackpad supports I2C mode with all the fancy gestures via this driver: https://github.com/leo-labs/VoodooSMBus
> > The tradeoff is the mylar coating which doesn't feel as nice as the glass trackpad on X1C6
> > I ordered the FRU on Aliexpress, will report back as soon as it arrives
> 
> If you want the fancy gestures, there's no need to buy another touchpad. Acidanthera's VoodooPS2 has full gestures on PS2 with V2.0.4 (2.1.0 is broken on X1C6). Am using it on my X1C6, gestures works exactly like a MacBook Pro.
> 
> [VoodooPS2](https://github.com/acidanthera/VoodooPS2/releases/download/2.0.4/VoodooPS2Controller-2.0.4-RELEASE.zip)

It's not the same at all. Definitely feels better than stock VoodooPS2, but it's very far from Macbook Pro, mostly due to the limitations of PS2.
For instance, there is no palm rejection and the option to ignore trackpad input while typing is missing (setting it in SSDT has no effect).

**Comment by AutinMitra** [comment]
> > > X1C6 has a Synaptics trackpad that only works in PS2 mode on macOS.
> > > However, after some digging it seems that ELAN trackpad from T480s should be physically compatible with X1C6. This trackpad supports I2C mode with all the fancy gestures via this driver: https://github.com/leo-labs/VoodooSMBus
> > > The tradeoff is the mylar coating which doesn't feel as nice as the glass trackpad on X1C6
> > > I ordered the FRU on Aliexpress, will report back as soon as it arrives
> > 
> > 
> > If you want the fancy gestures, there's no need to buy another touchpad. Acidanthera's VoodooPS2 has full gestures on PS2 with V2.0.4 (2.1.0 is broken on X1C6). Am using it on my X1C6, gestures works exactly like a MacBook Pro.
> > [VoodooPS2](https://github.com/acidanthera/VoodooPS2/releases/download/2.0.4/VoodooPS2Controller-2.0.4-RELEASE.zip)
> 
> It's not the same at all. Definitely feels better than stock VoodooPS2, but it's very far from Macbook Pro, mostly due to the limitations of PS2.
> For instance, there is no palm rejection and the option to ignore trackpad input while typing is missing (setting it in SSDT has no effect).

Ignore trackpad input is a default option in the kext, although the delay is too long (500ms), so I changed it to 75ms via the kext's plist. It's unfortunate that there's no palm rejection in either VoodooPS2 or VoodooI2C, but it's still great that the team has done so much to improve the quality! Let me know how the T480 trackpad works out!

**Comment by velaar** [comment]
@AutinMitra which FRU did you order? It seems there are quite a few fake touchpads on Aliexpress

**Comment by notthebee** [comment]
> @AutinMitra which FRU did you order? It seems there are quite a few fake touchpads on Aliexpress

I bought this one: https://www.aliexpress.com/item/32966730000.html?spm=a2g0s.9042311.0.0.23b34c4dSVGGfa

**Comment by tylernguyen** [comment]
> > @AutinMitra which FRU did you order? It seems there are quite a few fake touchpads on Aliexpress
> 
> I bought this one: https://www.aliexpress.com/item/32966730000.html?spm=a2g0s.9042311.0.0.23b34c4dSVGGfa

Hey @notthebee, how was the touchpad replacement process? And how's the experience with voodooi2c afterward?

**Comment by notthebee** [comment]
> > > @AutinMitra which FRU did you order? It seems there are quite a few fake touchpads on Aliexpress
> > 
> > 
> > I bought this one: https://www.aliexpress.com/item/32966730000.html?spm=a2g0s.9042311.0.0.23b34c4dSVGGfa
> 
> Hey @notthebee, how was the touchpad replacement process? And how's the experience with voodooi2c afterward?

Hey. I canceled my order since @leo-labs, the author of VoodooSMBUS mentioned that Synaptics touchpad is also connected via I2C and there might be a way to port the Linux driver to macOS https://github.com/leo-labs/VoodooSMBus/issues/21#issuecomment-563429735

Me and him are currently working on it.

**Comment by velaar** [comment]
@notthebee 

Can I ask you to keep us posted here? It is getting hard to follow all the X1C6 developments.

**Comment by notthebee** [comment]
> @notthebee
> 
> Can I ask you to keep us posted here? It is getting hard to follow all the X1C6 developments.

Will do!

**Comment by acoutts** [comment]
@sunrez did you ever start anything for the X1C7? I am looking at an X1 Carbon but would really prefer having the extra 2 cores. I'd be willing to help port things for the X1C7 but need some reading material for where to start.

Any help appreciated!

**Comment by sunrez** [comment]
Hi @acoutts ,

No, I held off on the X1C7 because unfortunately the WiFi is soldered and incompatible with macOS.   Based on Reddit threads people haven't had luck putting M.2 PCIe cards into the WWAN slot of X1C7 ; right now it looks like Lenovo only plumbed up USB2.0 to the WWAN connector in the X1C7 :( .  In the X1C6 they had PCIe going to the WWAN slot.

Since Apple has brought back the "classic" keyboard to the 16" MacPro would that be an option?  The high failure rate of the keyboard was a key reason I was looking at the X1C6.

Currently I use an XPS13/9360 but the coil whine is really awful.  If/when Apple brings the classic keyboard back to the 13" model I'll go back to that.

**Comment by acoutts** [comment]
Yeah I got my X1C7 setup and tried several things but in the end I decided I'll return it. 

I had trouble getting WiFi to work on Ubuntu 18.04 and had to use a custom back port driver. Then there were issues with pulse audio adjusting the wrong channels for audio (needed to modify it to only adjust PCM). Then the track pad - it just isn't even in the ballpark of macos and I tried 3 drivers (mtrack, synaptics, and libinput). But I found the X1C7 trackpad hardware can't detect palms or pressure so thumb detection wouldn't work.

In the end I gave up. Then I tried booting Clover but since there's no 10th gen Macs yet I had no luck even booting to the installer. It would hang at a black screen with the +'s. 

After all of this I realized even if I manage to boot it, WiFi and likely graphics won't work and the trackpad probably won't be very nice given it lacks the palm / pressure detection. 

I'm going to just keep dealing with my '14 13" MBP and wait for a refresh of the 13" line.

I do mobile development so I need macOS to compile iOS apps and run the iOS simulators. On my desktop I run Ubuntu and have macOS running in a kvm VM with hardware acceleration using VFIO to pass in a physical graphics card for native performance. Unfortunately this sort of thing just isn't possible on mobile yet. Even if we had a dGPU, they're all Nvidia and incompatible with macOS now.

Was a fun weekend of tinkering but disappointing for me in the end.

**Comment by sunrez** [comment]
Hi @acoutts,

Sorry to hear the X1C7 didn't work out :( .  If your ears aren't sensitive, the XPS13/9360/i7-8550u runs Mojave very nicely.  There's a repo that makes the operation fairly plug and play: https://github.com/the-darkvoid/XPS9360-macOS

My lingering issue is USB-C to HDMI doesn't work with audio, the coil whine and I wish the touchpad had palm rejection.

I'm still limping along with 2015 MBP 13" as well while I try to get the XPS13 going.

**Comment by acoutts** [comment]
@sunrez  before I mail this X1C7 back do you think it's worth trying one more time for wifi? I have here now a Dell BCM94350ZAE chip I was going to try in the WWAN slot, and I also have the WWAN antennas from Lenovo I was going to try plugging into it. But before I was stuck at even booting to the macos installer. Is there no hope until a 10th gen Intel Mac is released?

It's possible I didn't have my Clover setup with the right kernel extensions but I'm very new to this still. I also have a Samsung NVMe SSD here to try as your guide mentioned the one that ships with the X1C6 was incompatible (my X1C7 came with a Toshiba one). In my tests so far I had the stock HD in there. After I tried to boot the installer partition from clover it would hang on a black screen with some + signs but no other output to see in verbose mode. 

Also, based on your experience is the trackpad in macOS as good as it is in windows on these Lenovo laptops? I'm still on the fence if it's worth trying some more!

**Comment by sunrez** [comment]
Hi @acoutts,

If you have the Dell card I think it's definitely worth a shot putting it into the WWAN slot!  

After putting the card into the slot, enabling whatever other BIOS options are available, boot sysreccd Linux (found here: www.system-rescue-cd.org ), and see what lspci gives you.  If the card is detected, lspci should see it . If you can also grab the output of dmesg after the system is booted that will help too.  If you can put that up on pastbin.com or paste it here, that will be great.

If the WiFi problem can be worked out by putting a card in the WWAN slot, I'm personally game to get an X1C7 and work with you.

If the trackpad on the X1C7 is as good as the one in my Dell XPS13 it's IMO almost as good as a Mac.  The only thing the XPS13 touchpad is lacking is palm rejection which is fixable in software and I'm personally willing to give that a go anyway.

I replaced the Toshiba nVME SSD with a Samsung M.2 960 EVO SATA one and that worked perfectly for me, that's an easy fix if your issue is the SSD.

**Comment by acoutts** [comment]
Awesome I'm going to give it another go then. I just realized I need an NGFF adapter for this card to fit so that will take a few days to ship. Will keep you updated with it.

**Comment by acoutts** [comment]
Here's an update on the WWAN antenna in the meantime. I ordered part #5A30V25489 (https://encompass.com/item/12512960/LENOVO/5A30V25489/) and it seems to fit in some ways but not others.

Here's an overhead view of the X1C7 (10th gen i7) opened up. You can see the antennas at the bottom and they're shaped as if they should go in the bottom left/right corners, but there's nothing in the chassis to screw them down to (no screw holes anywhere).
![IMG_20191231_062050](https://user-images.githubusercontent.com/9597207/71620325-6010ef00-2b97-11ea-9ce2-0302f217e7da.jpg)

Here I've just set them in place and they seem to fit naturally in these spots (and the cables line up to where they should go on the board).
![IMG_20191231_062539](https://user-images.githubusercontent.com/9597207/71620350-8fbff700-2b97-11ea-948e-79aa1dd86fc3.jpg)

Close up of left side.
![IMG_20191231_062605](https://user-images.githubusercontent.com/9597207/71620373-a36b5d80-2b97-11ea-8eff-1a56c0056e51.jpg)

Close up of left side.
![IMG_20191231_062612](https://user-images.githubusercontent.com/9597207/71620377-acf4c580-2b97-11ea-80bb-0e9c104c94a6.jpg)

Close up of right side.
![IMG_20191231_062617](https://user-images.githubusercontent.com/9597207/71620398-c990fd80-2b97-11ea-8ad7-ef49fbc001b3.jpg)

Close up of right side.
![Uploading IMG_20191231_062624.jpg…]()

Sticking up here is the 5-pin connector on the right side antenna. There is a spot to plug it into on the board (just to the left of it here).
![IMG_20191231_062945](https://user-images.githubusercontent.com/9597207/71620409-dada0a00-2b97-11ea-8596-2a5741f3d532.jpg)

1) Will it affect things if the antenna aren't screwed into place? I'm not sure if there's some grounding happening with that. I'm sure they will hold in place fine with just the case on or maybe something to hold it there like silicon paste.
2) Do we need the 5 pin connector if all we want to do is use these as the WiFi antenna? The lenovo site lists the part is `Antenna Kits Wl+ww Amp` so that makes me think this is some kind of amp for the antenna but may not be needed for wifi.

I tested and the antenna cables do snap into place on the Dell wifi card.

**Comment by acoutts** [comment]
I started this repo to track my progress:
https://github.com/acoutts/x1c7-hackintosh

I have in there my stock DSDT files and so far a minimum working clover config where I'm able to get to the Clover bootloader, but so far unable to boot to macOS installer.

I had to steal the drivers folder from this repository to get to where I am now:
https://github.com/daliansky/XiaoXinPro-13-2019-hackintosh

So far I'm stuck here. It loads the kexts and then pauses on the last one where it says it's waiting for 5s:
```
83:598  0:057  SetStartupDiskVolume:
83:631  0:032    * Volume: 'Install macOS Catalina'
83:664  0:032    * LoaderPath: '\System\Library\CoreServices\boot.efi'
83:704  0:039    * DevPath: Install macOS Catalina
83:739  0:034    * GUID = E4388F94-FAD0-4120-A5A6-6E2A88C938CF
83:771  0:032    * efi-boot-device: <array><dict><key>IOMatch</key><dict><key>IOProviderClass</key><string>IOMedia</string><key>IOPropertyMatch</key><dict><key>UUID</key><string>E4388F94-FAD0-4120-A5A6-6E2A88C938CF</string></dict></dict></dict></array>
83:839  0:068  Custom boot screen not used because entry has unset use graphics
83:872  0:032  Closing log
```

**Comment by sunrez** [comment]
@acoutts,

While I think getting the antenna stuff figured out will be relevant, seeing if the M.2 PCIe card is recognized in the WWAN slot is the biggest issue.  There are IPEX extension cables that can be used to connect to the existing WiFi antennas.

**Comment by sunrez** [comment]
@notthebee ,

Any progress or status updates?  If you need assistance on the hardware side debugging, let me know.  I did write an I2C touchpad driver that worked under Linux for the XPS13 when debugging something.

If anyone has schematics for the X1C6 or X1C7 please also follow up.

**Comment by AutinMitra** [comment]
> @notthebee ,
> 
> Any progress or status updates? If you need assistance on the hardware side debugging, let me know. I did write an I2C touchpad driver that worked under Linux for the XPS13 when debugging something.
> 
> If anyone has schematics for the X1C6 or X1C7 please also follow up.

I found some schematics after a quick search, not sure if they're correct though. 

https://bioshelp.com/lenovo-x1-carbon-6th-gen-lcfc-yoda-2-nm-b481-schematics-bios.html

**Comment by sunrez** [comment]
@autinmitra,

Thanks for those links!

The section on the TrackPad / pointing stick (page 54 / 55) is a bit unclear but at first glance it does look like I2C is at least going to the connector.  Still not sure if the touchpad itself supports I2C or if it needs to be reconfigured from PS2 mode to I2C mode.

Getting a trackpad from a Thinkpad that supports I2C only (as was suggested earlier) seems like a very sane way to go.

I'll try to get an X1C6 in the near future and start experimenting.

**Comment by sunrez** [comment]
If the schematics are right, looks like two I2C controllers are used, one for NFC, the second connected to TrackPad connector.

**Comment by leo-labs** [comment]
Arch wiki lists both `i2c_i801` (which is i801 smbus) and `rmi_smbus` https://wiki.archlinux.org/index.php/Lenovo_ThinkPad_X1_Carbon_(Gen_7)

**Comment by sunrez** [comment]
@leo-labs,

I wrote some low-level Linux touchpad code for XPS13/9360/i7-8550u which is pretty much same as the X1C6.

The I2C controller that my code attached to (to talk with the trackpad on the XPS13) ende dup being of of the two below.  If you dig deep into the Linux drivers, one needs to use the synopsys i2c_dw code to actually do the work.

00:15.0 Serial bus controller [0c80]: Intel Corporation Cannon Point-LP Serial IO I2C Controller #0 (rev 11)
00:15.1 Serial bus controller [0c80]: Intel Corporation Cannon Point-LP Serial IO I2C Controller #1 (rev 11)

**Comment by tylernguyen** [comment]
@leo-labs @notthebee @sunrez Are there any updates on this topic? I would love to help if possible.

**Comment by sunrez** [comment]
@tylernguyen,

I haven't had time to look at this recently.  I don't have an X1C6 to mess around with at the moment, need to acquire one.

I think there are Linux utilities that can list / query I2C devices which might be something to look at.  Under sysfs in Linux one can see what it lists for the two I2C controllers I mentioned (these are the actual I2C controllers built into the CPU).

**Comment by tylernguyen** [comment]
**Update:**

It seems that VoodooSMBus has not been developed on for a while now. Since getting I2C to work on the x1c6 would require a driver port from Linux and VoodooSMBus update to [support OpenCore](https://github.com/leo-labs/VoodooSMBus/issues/20), both of which are out of my hand and [would require a lot of effort](https://github.com/leo-labs/VoodooSMBus/issues/21#issuecomment-563429735) from @leo-labs. I will close this issue for now.

Furthermore, I am very satisfied with VoodooPS2 currently, especially with my SSDT-Keyboard patch. In addition, I think that it will be better for us to spend more time optimizing VoodooPS2 behavior on the x1c6, as VoodooPS2 is being actively developed by the acidanthera team.

Should I see VoodooSMBus being picked up again, I will re-open this issue for further discussion.

**Comment by leo-labs** [comment]
I consider `VoodooSMBus` feature complete at the moment, far from dead. It works perfect for all ELAN touchpads and trackpoints.
I haven't looked into OpenCore compatibility but it should not be rather easy to add, PRs are welcome, but not a priority at the moment. 

 I am not able to add synaptics support because I do not own an x1c6. However I still happily offer my help to anybody who wants to try. This should not be very hard, porting it from linux is straightforward.

**Comment by tylernguyen** [comment]
@leo-labs Your help is much needed, please contact me on Signal at (202)-644-9951.

**Comment by tylernguyen** [comment]
My initial testing of https://github.com/VoodooSMBus/VoodooRMI is working very well. Gestures are now working much better and more reliably than PS2. I'm going to open this issue back up for testing purposes. On the next repo update, I will switch from PS2 to RMI permanently.

@sunrez @velaar @leo-labs 
Please spend some time testing and let me know your experience. The kext is still in beta so let's find some bugs to report.

@leo-labs @1Revenger1 Thank you for your hard work!

**Comment by velaar** [comment]
@tylernguyen can't comment on stability yet but in terms of precision and overall feel the RMI kext is way better than PS2. Gestures also work in a more native way.

**Comment by tylernguyen** [comment]
@velaar Thanks for confirming. And yup, that's my experience as well. Gestures feel way more smooth and natural. If possible, please keep using it as your default input kext for a few days, it's really a stability question at this point, particular after sleep/wake. 

Also, try to play with force touch, I feel that there's room to improve in that area. Let me know if you find anything!

**Comment by benbender** [comment]
I second that, but I had some irresponsive touchpad/trackpad after boot from time to time. I think one needs to block some Apple-Kexts which might lead to race conditions while attaching to smbus/i2c.

My take on this atm is:
````
                <key>Block</key>
                <array>
                        <dict>
                                <key>Comment</key>
                                <string>Prevent Apple I2C kexts from attaching to I2C controllers</string>
                                <key>Enabled</key>
                                <true/>
                                <key>Identifier</key>
                                <string>com.apple.driver.AppleIntelLpssI2C</string>
                                <key>MaxKernel</key>
                                <string></string>
                                <key>MinKernel</key>
                                <string></string>
                        </dict>
                        <dict>
                                <key>Comment</key>
                                <string>Prevent Apple I2C kexts from attaching to I2C controllers</string>
                                <key>Enabled</key>
                                <true/>
                                <key>Identifier</key>
                                <string>com.apple.driver.AppleIntelLpssI2CController</string>
                                <key>MaxKernel</key>
                                <string></string>
                                <key>MinKernel</key>
                                <string></string>
                        </dict>
                        <dict>
                                <key>Comment</key>
                                <string>Use VoodooSMBUS</string>
                                <key>Enabled</key>
                                <true/>
                                <key>Identifier</key>
                                <string>com.apple.driver.AppleSMBusPCI</string>
                                <key>MaxKernel</key>
                                <string></string>
                                <key>MinKernel</key>
                                <string></string>
                        </dict>
                        <dict>
                                <key>Comment</key>
                                <string>Use VoodooSMBUS</string>
                                <key>Enabled</key>
                                <false/>
                                <key>Identifier</key>
                                <string>com.apple.driver.AppleSMBusController</string>
                                <key>MaxKernel</key>
                                <string></string>
                                <key>MinKernel</key>
                                <string></string>
                        </dict>
                </array>
````

**Comment by tylernguyen** [comment]
@benbender The dev team and I know that race condition exists. See https://github.com/VoodooSMBus/VoodooSMBus/pull/41
However, we suspect that the Apple kexts are not the cause of that problem.

You could join us at the VoodooSMBUS gitter room to discuss this issue. https://gitter.im/VoodooSMBus/dev

**Comment by benbender** [comment]
Oh, nice info. Thx for the hint. Now we just need a new release with the option to disable the Elan-driver... Maybe I'll patch it at the weekend until than. :)

**Comment by tylernguyen** [comment]
Implemented `VoodooRMI` as default touchpad input as of commit  a985053
For all bug reports, improvements, and issues concerning the kext, please do so at 
https://github.com/VoodooSMBus/VoodooRMI
